File: feature_tests/java/somelib/src/main/java/dev/diplomattest/somelib/Foo.java
package dev.diplomattest.somelib;

import dev.diplomattest.somelib.ntv.*;


import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.SegmentAllocator;
import java.lang.ref.Cleaner;
import java.util.List;
import static java.lang.foreign.ValueLayout.*;
import java.nio.charset.StandardCharsets;
import java.util.stream.Stream;

public class Foo {

    MemorySegment internal;
    Cleaner.Cleanable cleanable;
    SegmentAllocator arena;

    List<Object> selfEdges = List.of();
    List<Object> aEdges = List.of();
    

    static class FooCleaner implements Runnable {

        MemorySegment segment;
        FooCleaner(MemorySegment segment) {
            this.segment = segment;
        }

        public void run() {
            somelib_h.Foo_destroy(this.segment);
        }
    }

    Foo() {}
    Foo(MemorySegment handle, List<Object> selfEdges, List<Object> aEdges) {
        this.internal = handle;
        this.selfEdges = selfEdges;
        this.aEdges = aEdges;
        

    }
    
    public static Foo new_(String x) {
        
        try (var arena = Arena.ofConfined()) {
            var xData= Arena.ofAuto().allocateFrom(x, StandardCharsets.UTF_8);
            var xLen = xData.byteSize() - 1;  // allocated strings are null terminated
            var xView = DiplomatStringView.allocate(Arena.ofAuto());
            DiplomatStringView.len(xView, xLen);
            DiplomatStringView.data(xView, xData);
            var nativeVal = somelib_h.Foo_new(xView);
            
            List<Object> selfEdges = List.of();
            
            
            
            List<Object> aEdges = List.of(x);
            var returnVal = new Foo(nativeVal, selfEdges, aEdges);
            return returnVal;
                    
        }
    }
    
    public static Foo extractFromFields(BorrowedFields fields) {
        
        try (var arena = Arena.ofConfined()) {
            var fieldsNative = fields.toNative(arena);
            var nativeVal = somelib_h.Foo_extract_from_fields(fieldsNative);
            
            List<Object> selfEdges = List.of();
            
            
            
            List<Object> aEdges = List.of(fields);
            var returnVal = new Foo(nativeVal, selfEdges, aEdges);
            return returnVal;
                    
        }
    }
    
    public static Foo extractFromBounds(BorrowedFieldsWithBounds bounds,String anotherString) {
        
        try (var arena = Arena.ofConfined()) {
            var boundsNative = bounds.toNative(arena);
            var anotherStringData= Arena.ofAuto().allocateFrom(anotherString, StandardCharsets.UTF_8);
            var anotherStringLen = anotherStringData.byteSize() - 1;  // allocated strings are null terminated
            var anotherStringView = DiplomatStringView.allocate(Arena.ofAuto());
            DiplomatStringView.len(anotherStringView, anotherStringLen);
            DiplomatStringView.data(anotherStringView, anotherStringData);
            var nativeVal = somelib_h.Foo_extract_from_bounds(boundsNative, anotherStringView);
            
            List<Object> selfEdges = List.of();
            
            
            
            List<Object> aEdges = List.of(bounds, bounds, anotherString);
            var returnVal = new Foo(nativeVal, selfEdges, aEdges);
            return returnVal;
                    
        }
    }
    
    
    public Bar getBar() {
        
        
        var nativeVal = somelib_h.Foo_get_bar(internal);
        
        List<Object> selfEdges = List.of();
        
        
        
        List<Object> bEdges = List.of(this);
        List<Object> aEdges = List.of(this);
        var returnVal = new Bar(nativeVal, selfEdges, bEdges, aEdges);
        return returnVal;
                
    }
    
    public BorrowedFieldsReturning asReturning() {
        
        try (var arena = Arena.ofConfined()) {
            
            
            var nativeVal = somelib_h.Foo_as_returning(arena, internal);
            
            var returnVal = new BorrowedFieldsReturning(nativeVal, List.of());
            return returnVal;
                    
        }
    }
    
}

File: feature_tests/kotlin/somelib/src/main/kotlin/dev/diplomattest/somelib/Foo.kt
package dev.diplomattest.somelib;
import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure

internal interface FooLib: Library {
    fun Foo_destroy(handle: Pointer)
    fun Foo_new(x: Slice): Pointer
    fun Foo_get_bar(handle: Pointer): Pointer
    fun Foo_new_static(x: Slice): Pointer
    fun Foo_as_returning(handle: Pointer): BorrowedFieldsReturningNative
    fun Foo_extract_from_fields(fields: BorrowedFieldsNative): Pointer
    fun Foo_extract_from_bounds(bounds: BorrowedFieldsWithBoundsNative, anotherString: Slice): Pointer
}

class Foo internal constructor (
    internal val handle: Pointer,
    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>,
    internal val aEdges: List<Any?>,
)  {

    internal class FooCleaner(val handle: Pointer, val lib: FooLib) : Runnable {
        override fun run() {
            lib.Foo_destroy(handle)
        }
    }
    fun registerCleaner() {
        CLEANER.register(this, Foo.FooCleaner(handle, Foo.lib));
    }

    companion object {
        internal val libClass: Class<FooLib> = FooLib::class.java
        internal val lib: FooLib = Native.load("diplomat_feature_tests", libClass)
        @JvmStatic
        
        fun new_(x: String): Foo {
            // This lifetime edge depends on lifetimes: 'a
            val aEdges: MutableList<Any> = mutableListOf();
            val xSliceMemory = PrimitiveArrayTools.borrowUtf8(x).into(listOf(aEdges))
            
            val returnVal = lib.Foo_new(xSliceMemory.slice);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = Foo(handle, selfEdges, aEdges)
            returnOpaque.registerCleaner()
            return returnOpaque
        }
        @JvmStatic
        
        fun newStatic(x: String): Foo {
            // This lifetime edge depends on lifetimes: 'a
            val aEdges: MutableList<Any> = mutableListOf();
            val xSliceMemory = PrimitiveArrayTools.borrowUtf8(x).leakStatic()
            
            val returnVal = lib.Foo_new_static(xSliceMemory.slice);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = Foo(handle, selfEdges, aEdges)
            returnOpaque.registerCleaner()
            return returnOpaque
        }
        @JvmStatic
        
        fun extractFromFields(fields: BorrowedFields): Foo {
            // This lifetime edge depends on lifetimes: 'a
            val aEdges: MutableList<Any> = mutableListOf();
            
            val returnVal = lib.Foo_extract_from_fields(fields.toNative(aAppendArray = arrayOf(aEdges)));
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = Foo(handle, selfEdges, aEdges)
            returnOpaque.registerCleaner()
            return returnOpaque
        }
        @JvmStatic
        
        /** Test that the extraction logic correctly pins the right fields
        */
        fun extractFromBounds(bounds: BorrowedFieldsWithBounds, anotherString: String): Foo {
            val temporaryEdgeArena: MutableList<Any> = mutableListOf()
            // This lifetime edge depends on lifetimes: 'a, 'y, 'z
            val aEdges: MutableList<Any> = mutableListOf();
            val anotherStringSliceMemory = PrimitiveArrayTools.borrowUtf8(anotherString).into(listOf(aEdges))
            
            val returnVal = lib.Foo_extract_from_bounds(bounds.toNative(aAppendArray = arrayOf(temporaryEdgeArena), bAppendArray = arrayOf(aEdges), cAppendArray = arrayOf(aEdges)), anotherStringSliceMemory.slice);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = Foo(handle, selfEdges, aEdges)
            returnOpaque.registerCleaner()
            return returnOpaque
        }
    }
    
    fun getBar(): Bar {
        // This lifetime edge depends on lifetimes: 'a
        val aEdges: MutableList<Any> = mutableListOf(this);
        // This lifetime edge depends on lifetimes: 'a, 'b
        val bEdges: MutableList<Any> = mutableListOf(this);
        
        val returnVal = lib.Foo_get_bar(handle);
        val selfEdges: List<Any> = listOf()
        val handle = returnVal 
        val returnOpaque = Bar(handle, selfEdges, bEdges, aEdges)
        returnOpaque.registerCleaner()
        return returnOpaque
    }
    
    fun asReturning(): BorrowedFieldsReturning {
        // This lifetime edge depends on lifetimes: 'a
        val aEdges: MutableList<Any> = mutableListOf(this);
        
        val returnVal = lib.Foo_as_returning(handle);
        val returnStruct = BorrowedFieldsReturning.fromNative(returnVal, aEdges)
        return returnStruct
    }

}